test/metabase/query_processor/pivot_test.clj

```diff
@@ -15,54 +16,167 @@
-                                                 [:expression "pivot-grouping"]]
-                                   :expressions {"pivot-grouping" [:abs 5]}}}
-
-                          {:query {:breakout    [[:expression "pivot-grouping"]]
-                                   :expressions {"pivot-grouping" [:abs 7]}}}]
+                            {:query {:breakout    [[:expression "pivot-grouping"]]
+                                     :expressions {"pivot-grouping" [:abs 7]}}}])
                 expected (map (fn [expected-val] (-> expected-val
                                                      (assoc :type       :query
                                                             :parameters []
-                                                            :pivot_rows [1 0]
-                                                            :pivot_cols [2])
+                                                            :pivot-rows [1 0]
+                                                            :pivot-cols [2])
                                                      (assoc-in [:query :fields] [[:expression "pivot-grouping"]])
                                                      (assoc-in [:query :aggregation] [[:count] [:sum (mt/$ids $orders.quantity)]])
                                                      (assoc-in [:query :source-table] (mt/$ids $$orders)))) expected)
-                actual   (map (fn [actual-val] (dissoc actual-val :database)) (sut/generate-queries request))]
-            (is (= 7 (count actual)))
+                actual   (map (fn [actual-val] (dissoc actual-val :database)) (#'pivot/generate-queries request))]
+            (is (= 5 (count actual)))
+            (is (= (map #(get-in % [:query :expressions "pivot-grouping"]) expected)
+                   (map #(get-in % [:query :expressions "pivot-grouping"]) actual)))
             (is (= expected actual))))))))
+
+(deftest breakout-combinations-test
+  (testing "Should return the combos that Paul specified in (#14329)"
+    (is (= [[0 1 2]
+            [0 1  ]
+            [0    ]
+            [     ]]
+           (#'pivot/breakout-combinations 3 [0 1 2] [])))
+    (is (= (sort-by
+            (partial #'pivot/group-number 4)
+            [ ;; subtotal rows
+             [0   3]
+             [0 1 3]
+             ;; row totals
+             [0 1 2]
+             ;; subtotal rows within "row totals"
+             [0    ]
+             [0 1  ]
+             ;; "grand totals" row
+             [    3]
+             ;; bottom right corner
+             [     ]])
+           (#'pivot/breakout-combinations 4 [0 1 2] [3])))))
+
+(deftest group-number-test
+  (doseq [[indecies expected] {[0]     6
+                               [0 1]   4
+                               [0 1 2] 0
+                               []      7}]
+    (is (= expected
+           (#'pivot/group-number 3 indecies)))))
```

Comment:
I'm not sure exactly why the group numbers are determined this way but I made sure to match existing behavior. I was considering changing to group *names* like `001` or `010` but didn't want to break the FE

Reply:
That would break the frontend. See above comment about the bitmask calculation.